engine/src/main/java/org/hibernate/validator/internal/xml/MappingXmlParser.java

```diff
@@ -109,105 +90,44 @@ public MappingXmlParser(ConstraintHelper constraintHelper, TypeResolutionHelper
-					classLoadingHelper,
-					constraintHelper,
-					typeResolutionHelper,
-					valueExtractorManager
-			);
-			GroupConversionBuilder groupConversionBuilder = new GroupConversionBuilder( classLoadingHelper );
-
-			ConstrainedTypeBuilder constrainedTypeBuilder = new ConstrainedTypeBuilder(
-					classLoadingHelper,
-					metaConstraintBuilder,
-					annotationProcessingOptions,
-					defaultSequences
-			);
-			ConstrainedFieldBuilder constrainedFieldBuilder = new ConstrainedFieldBuilder(
-					metaConstraintBuilder,
-					groupConversionBuilder,
-					annotationProcessingOptions
-			);
-			ConstrainedExecutableBuilder constrainedExecutableBuilder = new ConstrainedExecutableBuilder(
-					classLoadingHelper,
-					metaConstraintBuilder,
-					groupConversionBuilder,
-					annotationProcessingOptions
-			);
-			ConstrainedGetterBuilder constrainedGetterBuilder = new ConstrainedGetterBuilder(
-					metaConstraintBuilder,
-					groupConversionBuilder,
-					annotationProcessingOptions
-			);
-
 			Set<String> alreadyProcessedConstraintDefinitions = newHashSet();
 			for ( InputStream in : mappingStreams ) {
-				ConstraintMappingsType mapping = unmarshal( jc, in );
-				String defaultPackage = mapping.getDefaultPackage();
-
-				parseConstraintDefinitions(
-						mapping.getConstraintDefinition(),
-						defaultPackage,
-						alreadyProcessedConstraintDefinitions
-				);
+				// the InputStreams passed in parameters support mark and reset
+				in.mark( Integer.MAX_VALUE );
 
-				for ( BeanType bean : mapping.getBean() ) {
-					processBeanType(
-							constrainedTypeBuilder,
-							constrainedFieldBuilder,
-							constrainedExecutableBuilder,
-							constrainedGetterBuilder,
-							defaultPackage,
-							bean
-					);
-				}
+				XMLEventReader xmlEventReader = xmlParserHelper.createXmlEventReader( "constraint mapping file", new CloseIgnoringInputStream( in ) );
+				String schemaVersion = xmlParserHelper.getSchemaVersion( "constraint mapping file", xmlEventReader );
+				xmlEventReader.close();
 
 				in.reset();
-			}
-		}
-		catch (JAXBException | SAXException | IOException | XMLStreamException e) {
-			throw LOG.getErrorParsingMappingFileException( e );
-		}
-	}
 
-	private ConstraintMappingsType unmarshal(JAXBContext jc, InputStream in) throws JAXBException, XMLStreamException, IOException, SAXException {
-		ClassLoader previousTccl = run( GetClassLoader.fromContext() );
+				// The validation is done first as we manipulate the XML document before pushing it to the unmarshaller
+				// and it might not be valid anymore as we might have switched the namespace to the latest namespace
+				// supported.
+				String schemaResourceName = getSchemaResourceName( schemaVersion );
+				Schema schema = xmlParserHelper.getSchema( schemaResourceName );
+				Validator validator = schema.newValidator();
+				validator.validate( new StreamSource( new CloseIgnoringInputStream( in ) ) );
```

Comment:
I'm wondering if we shouldn't validate the namespace somehow. It was done before by JAXB and now we don't check anything at all meaning that the user could have a random XML file there and we wouldn't throw any error, it would just be ignored.

Am I right?

Reply:
Wouldn't this be caught as part of the schema validation?

Reply:
yes, I would think that `Validator` should handle such case. I'll try that and add a test with some random xml.

Reply:
Yeah, sorry, don't know what I was thinking :).

Reply:
@marko-bekhta no need for an additional test here by the way.